protocols/decentralizedprotocol.py

```diff
@@ -0,0 +1,68 @@
+from absl import logging
+import numpy as np
+from random import randint
+
+from bandits import ordinarybandit
+from .utils import Protocol
+
+__all__ = ['DecentralizedProtocol']
+
+
+class DecentralizedProtocol(Protocol):
+  """Decentralized protocol
+  """
+
+  def __init__(self, pars):
+    self.__messages = []
+    self.__num_players = pars['num_players']
+
+  @property
+  def type(self):
+    return 'decentralizedprotocol_'+self._player_type
+
+  @property
+  def _num_players(self):
+    return self.__num_players
+
+  @property
+  def _messages(self):
+    return self.__messages
+
+  def _broadcast_message(self, message, player):
+    self.__messages.append(message)
+
+  def __init(self):
+    # time starts from 1
+    for k in range(self._num_players):
+      player = self._players[k]
+      bandit = self._bandits[k]
+      player.init(bandit)
+
+  def _one_trial(self, seed):
+    np.random.seed(seed)
+    ############################################################################
+    # initialization
+    self.__init()
+    ############################################################################
+    agg_regret = dict()
+
+    for t in range(self._horizon + 1):
+      if t > 0:
+        self._play_round(t)
+      if t > 0 and t % self._frequency == 0:
+        agg_regret[t] = self.regret()
+    return dict({self.type: agg_regret})
+
+  def _play_round(self, t):
+    # sample player
```

Comment:
I did not get it. I am assuming we should play each player once during each iteration?

Reply:
Right - I can switch it to that for regret minimization, or I can add a distributed protocol doing that as well.

I was thinking about the setting from here: [Feraud et al., 2019](https://arxiv.org/abs/1811.07763) which samples players according to a fixed distribution each iteration for arm identification.

Reply:
Okay, it's fine as long as you know what is going on. We should be careful about the protocol names. Anyway, feel free to give the names you think is good. We can change it later.

Reply:
Okay - sounds good. My plan is to get familiar asap so I can focus on decentralized arm selection. I definitely don't want to spoil your codebase since it was really clean before I started making commits.